### PR TITLE
【generate】add csa/hca kv cache generate

### PR DESCRIPTION
**Pull request by xiaoguoguo626807**
### PR Category

Execute Infrastructure

### PR Types

Improvements

### Description

支持csa attention 的kv cache 存储和读取逻辑

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py [hunk 1]
```diff
@@ -274,6 +279,13 @@ def forward(
         _, sq, _ = hidden_states.shape
         position_offset = cp_rank * sq if cp_size > 1 else 0
 
+        # Incremental decode: the new token's absolute position is the number
+        # of raw tokens already cached; RoPE (forward + inverse) must use it.
+        if use_cache and past_key_values is not None and cp_size == 1:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsv4_hybrid_attention.py:794]: P1 这里用统一的 raw cache 长度作为 RoPE 位置，会让左填充 batch 中不同样本的解码位置都变成同一个 `prompt_len`。`GreedyGenerator` 已按 `attention_mask` 生成逐样本 `position_ids`，但 DSv4/HCA 路径没有使用这些位置且 pad token 仍进入 raw/compressed cache，左填充生成会静默出错；请接入逐样本位置/填充 mask，或显式拒绝带 mask 的该路径。

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 2]
```diff
@@ -433,12 +433,34 @@ def forward(
...
                 is_causal = query.shape[1] > 1
                 if query.shape[1] == 1:
-                    attn_mask_kv = None
+                    if attention_mask is not None:
+                        # attention_mask shape: broadcastable to
+                        # [bsz, num_heads, q_len_prefill, kv_len_prefill]; take the
+                        # last query row (bool mask over key positions that are
+                        # valid, i.e. not padding) and broadcast it to the current
+                        # (possibly longer) KV length for this decode step.
+                        pad_mask = attention_mask[:, :, -1:, :]
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py]: P1 这里仍按 4-D mask 取 `attention_mask[:, :, -1:, :]`，但 `GreedyGenerator.generate` 新接口允许传入 `[B, L]`。该 2-D mask 会从 embedding 原样传到这里，解码步会直接索引失败；请先规范成 `[B, 1, 1, L]` 再截取/扩展。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现两处生成缓存路径的问题需要先修复，见行内评论；另外请补充 PR 描述中的变更说明和测试情况。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by risemeup1111** [comment]
### Nyanpasu Review Board


| 序号 | 位置 | 优先级 | 状态 |
| - | - | - | - |
| 1 | 非行级：当前分支仍有合并冲突 | P0 | ✅ |
| 2 | [`src/paddlefleet/transformer/dsv4_hybrid_attention.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1575#discussion_r3679511288) | P1 | ✅ |
| 3 | [`src/paddlefleet/transformer/dot_product_attention.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1575#discussion_r3679511464) | P1 | ✅ |
| 4 | 非行级：PR 描述缺少变更说明/测试计划 | P3 | ✅ |
| 5 | [`src/paddlefleet/generation/greedy_generator.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1575#discussion_r3681234046) | P2 | 🟡 |
| 6 | [`src/paddlefleet/transformer/csa_attention.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1575#discussion_r3689307164) | P2 | ✅ |
| 7 | [`src/paddlefleet/generation/csa_cache.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1575#discussion_r3689308418) | P3 | ✅ |
| 8 | [`src/paddlefleet/transformer/csa_attention.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1575#discussion_r3700962874) | P0 | ✅ |
| 9 | [`src/paddlefleet/transformer/csa_attention.py`](https://github.com/PaddlePaddle/PaddleFleet/pull/1575#discussion_r3700963858) | P1 | ✅ |

<details>
<summary>点击展开 Review 规则</summary>
PR 评审规则：

- P0、P1 级别的评审必须提交新的 commit 进行修改；
- P2、P3 级别的评审可以通过评论进行修改。

状态说明：

| 已解决 | 无需修复 | 待修复 |
| - | - | - |
| ✅ | 🟡 | 🚧 |
</details>

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/generation/greedy_generator.py [hunk 3]
```diff
@@ -457,7 +478,6 @@ def generate(
         """
         if input_ids.ndim != 2:
```
Comment (risemeup1111) [inline src/paddlefleet/generation/greedy_generator.py:479]: P2 这里缺少 `max_new_tokens <= 0` 的短路，cache 路径会先 prefill 并采样 1 个 token；而 `no_cache=True` 会因循环次数为 0 直接返回原 prompt，两个分支在边界值下行为不一致。请在进入 cache prefill 前直接返回原 prompt（`return_log_probs=True` 时返回空列表）。

**Review by risemeup1111 — COMMENTED** [review]
此前两处 P1 已验证修复；本轮仅剩一个 `max_new_tokens=0` 的边界问题，见行内评论。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/generation/greedy_generator.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/generation/greedy_generator.py:479]: 当前提交只新增了 `max_new_tokens < 0` 校验，`max_new_tokens == 0` 仍会继续走 cache prefill 并采样首个 token；这个边界问题还未解决。

File: src/paddlefleet/generation/greedy_generator.py [hunk 4]
```diff
@@ -457,7 +478,8 @@ def generate(
...
             raise ValueError("input_ids must be [B, L]")
-
+        if max_new_tokens < 0:
```
Comment (xiaoguoguo626807) [inline src/paddlefleet/generation/greedy_generator.py]:
```suggestion
        if max_new_tokens <= 0:
```

File: src/paddlefleet/generation/greedy_generator.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/generation/greedy_generator.py:479]: 这个提交已避免 `max_new_tokens == 0` 时继续采样，但现在会直接抛 `ValueError`，且文案仍写 `>= 0`。如果 0 要作为非法值，请同步改成 `> 0` 的文案/文档；否则仍建议按原评论在这里短路返回 prompt（log_probs 为空）。

Comment (xiaoguoguo626807) [reply src/paddlefleet/generation/greedy_generator.py:479]: 这个配置本身没有意义，就应该报错

Comment (risemeup1111) [reply src/paddlefleet/generation/greedy_generator.py:479]: 接受这个设计取舍。当前代码已在进入 cache/no_cache 分支前拒绝 `max_new_tokens == 0`，原先的分支行为不一致问题按已答复处理。

**Review by risemeup1111 — APPROVED** [review]

File: src/paddlefleet/transformer/csa_attention.py [hunk 5]
```diff
@@ -2223,12 +2420,30 @@ def forward(
...
+            layer_idx: this layer's index into ``past_key_values``.
+            use_cache: enable KV-cache prefill priming / incremental decode.
 
         Returns:
             output: [b, sq, np * v_head_dim]
         """
         b, sq, np_heads, hn = query.shape
 
+        # Incremental decode: single-token step against the cached state.
+        if use_cache and past_key_values is not None and sq == 1:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/csa_attention.py:2631]: P2 这里的 decode 早退只检查了 `attention_mask`，但 `DSv4HybridSelfAttention.forward()` 仍会把 `docmask_meta` 传进来；只要上游带着 `attn_mask_startend_row_indices` 走 cache，packed-doc 边界就会被静默忽略。请在这里一并拒绝 `docmask_meta`，或者把它继续传到 `_forward_decode`。

File: src/paddlefleet/generation/csa_cache.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        if self.k is None:
+            self.k = k_new
+            self.v = v_new
+        else:
+            self.k = paddle.concat([self.k, k_new], axis=1)
+            self.v = paddle.concat([self.v, v_new], axis=1)
+        return self.k, self.v
+
+
+class CSADynamicCache:
```
Comment (risemeup1111) [inline src/paddlefleet/generation/csa_cache.py:124]: P3 这个 cache 目前没有 `_seq_len` 兼容字段，但 `GreedyGenerator` 的 `_DEBUG` 分支还会直接读它；DSv4-Hybrid 一旦打开调试日志就会在 prefill 阶段 `AttributeError`。建议补一层兼容字段，或者把调试探针改成 `get_seq_len()`。

**Review by risemeup1111 — COMMENTED** [review]
当前实现还有两个边界点需要补齐，先修正后我再复核。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复核最新提交，前两处边界问题已补齐，当前未见新的阻塞性问题。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/csa_attention.py [hunk 7]
```diff
@@ -2572,8 +2791,124 @@ def forward(
...
+                compress_idxs = get_compress_topk_idxs_decode(b, offset, n_comp)
+            if compress_idxs.dtype != window_idxs.dtype:
+                compress_idxs = compress_idxs.cast(window_idxs.dtype)
+            topk_idxs = paddle.concat([window_idxs, compress_idxs], axis=-1)
+        else:
+            topk_idxs = window_idxs
+
+        topk_idxs = topk_idxs.cast("int32")
+
+        # 6. Sparse attention.
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/csa_attention.py:2946]: P0 `_forward_decode()` 在这里构造完 `topk_idxs` 后没有调用 `compressed_sparse_attn` 也没有返回值；`DSv4HybridSelfAttention._full_attn_forward()` 随后会访问 `core_attn_out.shape`，所以任意 `use_cache=True` 的单 token decode 都会拿到 `None` 并中断生成。

File: src/paddlefleet/transformer/csa_attention.py [hunk 8]
```diff
@@ -2400,12 +2597,34 @@ def forward(
...
+            assert attention_mask is None, (
+                "CSA incremental decode does not support attention_mask."
+            )
+            assert docmask_meta is None, (
+                "CSA incremental decode does not support packed documents; "
+                "pass attn_mask_startend_row_indices=None when decoding with "
+                "a cache."
+            )
+            state = past_key_values.get_csa_state(layer_idx)
+            return self._forward_decode(query, key, x, qr, state)
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/csa_attention.py:2644]: P1 这个 decode 早退会先于 `self.is_mqa_layer` 分支匹配 `compress_ratio=-1`，而 MQA prefill 在下面 `_forward_mqa()` 直接返回、不会填充 cache；因此 MQA cache decode 既丢历史 KV，也不会使用 full-causal topk。请为 MQA 单独处理 cache 或显式拒绝 `use_cache`。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/csa_attention.py (hunk 8 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/csa_attention.py:2644]: 当前提交已补回 `_forward_decode()` 的 return，但 MQA cache 问题仍在：`sq == 1` 时仍先进入通用 `_forward_decode()`，而 `_forward_mqa()` 的 prefill 路径仍直接返回、没有为 MQA 写入历史 KV。该 P1 仍需要代码修复。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — APPROVED** [review]
已复核当前提交，MQA cache 的 prefill 与单 token decode 路径已补齐，未见新的阻塞问题。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]

**Review by changeyoung98 — APPROVED** [review]
LGTM
